src/main/scala/org/ergoplatform/nodeView/state/ErgoState.scala

```diff
@@ -76,47 +77,55 @@ object ErgoState extends ScorexLogging {
 
   /**
     * Tries to validate and execute transactions.
-    *
+    * @param transactions to be validated and executed
+    * @param currentStateContext to be used for tx execution
+    * @param checkBoxExistence function to provide ErgoBox by BoxId
     * @return Result of transactions execution with total cost inside
     */
   def execTransactions(transactions: Seq[ErgoTransaction],
                        currentStateContext: ErgoStateContext)
                       (checkBoxExistence: ErgoBox.BoxId => Try[ErgoBox]): ValidationResult[Long] = {
-    import cats.implicits._
     val verifier: ErgoInterpreter = ErgoInterpreter(currentStateContext.currentParameters)
 
     @tailrec
-    def execTx(txs: List[ErgoTransaction], accCostTry: ValidationResult[Long]): ValidationResult[Long] = (txs, accCostTry) match {
-      case (tx :: tail, r: Valid[Long]) =>
-        val boxesToSpendTry: Try[List[ErgoBox]] = tx.inputs.toList
-          .map(in => checkBoxExistence(in.boxId))
-          .sequence
-
-        lazy val dataBoxesTry: Try[List[ErgoBox]] = tx.dataInputs.toList
-          .map(in => checkBoxExistence(in.boxId))
-          .sequence
-
-        lazy val boxes: Try[(List[ErgoBox], List[ErgoBox])] = dataBoxesTry.flatMap(db => boxesToSpendTry.map(bs => (db, bs)))
-
-        val vs = tx.validateStateless()
-          .validateNoFailure(txBoxesToSpend, boxesToSpendTry)
-          .validateNoFailure(txDataBoxes, dataBoxesTry)
-          .payload[Long](r.value)
-          .validateTry(boxes, e => ModifierValidator.fatal("Missed data boxes", e)) { case (_, (dataBoxes, toSpend)) =>
-            tx.validateStateful(toSpend.toIndexedSeq, dataBoxes.toIndexedSeq, currentStateContext, r.value)(verifier).result
-          }
-
-        execTx(tail, vs)
-      case _ =>
-        accCostTry
+    def collectBoxesById(
+                 remainingBoxIds: Iterator[ErgoBox.BoxId],
+                 resultingBoxes: Try[VectorBuilder[ErgoBox]] = Success(new VectorBuilder[ErgoBox])
+               ): Try[IndexedSeq[ErgoBox]] = {
+      if (!remainingBoxIds.hasNext) {
+        resultingBoxes.map(_.result())
+      } else {
+        checkBoxExistence(remainingBoxIds.next()) match {
+          case Success(box) =>
+            collectBoxesById(remainingBoxIds, resultingBoxes.map(_ += box))
```

Comment:
ArrayBuilder will be much faster than VectorBuilder, especially in this use case, where you can use arrBuilder.sizeHint(tx.inputs.length) to do single allocations. Then appending with += will be super fast.
Just compare what happens on += in VectorBuilder and ArrayBuilder.